packages/server/src/api/routes/tests/row.spec.ts

```diff
@@ -608,6 +618,32 @@ describe.each([
       await assertRowUsage(rowUsage)
     })
 
+    it("should update only the fields that are supplied and emit the correct oldRow", async () => {
+      let beforeRow = await config.api.row.save(table._id!, {
+        name: "test",
+        description: "test",
+      })
+
+      const rowUpdateEventPromise = waitForRowUpdateEvent()
+
+      const row = await config.api.row.patch(table._id!, {
+        _id: beforeRow._id!,
+        _rev: beforeRow._rev!,
+        tableId: table._id!,
+        name: "Updated Name",
+      })
+
+      expect(row.name).toEqual("Updated Name")
+      expect(row.description).toEqual(beforeRow.description)
+
+      const event = await rowUpdateEventPromise
```

Comment:
Usually with stuff like this where you have some beginning and then some end and something happens in between I like to wrap it in a callback function like this:

```typescript
const event = await waitForEvent("row:update", async () => {
  const row = await config.api.row.patch(table._id!, {
    _id: beforeRow._id!,
    _rev: beforeRow._rev!,
    tableId: table._id!,
    name: "Updated Name",
  })

  expect(row.name).toEqual("Updated Name")
  expect(row.description).toEqual(beforeRow.description)
})
```

Leaves a little bit less room for mistakes.

```typescript
function waitForEvent<T extends Event>(name: string, callback: () => Promise<void>): Promise<T> {
  const p = new Promise((resolve: any) => emitter.once(name, resolve))
  await callback()
  return p
}
```

Thinking a bit more about this, do we have to be careful that the event emitter is a global thing? So it could be that if tests are running in parallel the event that we get back in here is not the event from the test that we're running, it's from some other test.

Reply:
> do we have to be careful that the event emitter is a global thing

I think it is in this instance yea - not sure how to handle this exactly. Do you think something like instantiating a new emitter for this specific test would work better? Or what way should we handle this?

Reply:
Okay, I think I have a solution.

```typescript
interface WaitOptions {
  name: string,
  matchFn?: (event: any) => boolean,
}
async function waitForEvent(opts: WaitOptions, callback: () => Promise<void>): Promise<any> {
  const p = new Promise((resolve: any) => {
    const listener = (event) => {
      if (opts.matchFn && !opts.matchFn(event)) {
        return
      }
      resolve(event)
      emitter.off(name, listener)
    };
    emitter.on(name, listener)
  })
}
```

This gives you the ability to listen to an event with a function that allows you to filter which event you get back So what we can then do is use this function to match a specific row ID, like this:

```typescript
const opts = { name: "row:updated", matchFn: (event) => event.row._id === row._id }
const event = waitForEvent(opts, () => {
  // do stuff
})
```

I'm not fully sold on the API I've created there, but you see the idea. This gets around the problem of it being global, because you're looking for an event that you know that only you're going to match.